File: strands-py/src/strands/vended_interventions/cedar/_schema_generator.py
"""Adapter for cedar-policy-mcp-schema-generator package."""

from __future__ import annotations

import re
from typing import Any, cast

from typing_extensions import Required, TypedDict

DEFAULT_STUB = """
namespace Agent {
  @mcp_principal
  entity User;
  @mcp_resource
  entity Resource;
  @mcp_context("session")
  type SessionContext = {
    hour_utc: Long,
    call_count: Long
  };
}
"""


class ToolDefinition(TypedDict, total=False):
    """Minimal tool definition for schema generation (MCP-compatible format)."""

    name: Required[str]
    inputSchema: dict[str, Any]
    description: str


def generate_cedar_schema(tools: list[ToolDefinition]) -> str:
    """Generate a Cedar schema from tool definitions using cedar-policy-mcp-schema-generator.

    Args:
        tools: List of MCP-format tool definitions.

    Returns:
        Cedar schema text with namespace wrapper stripped.

    Raises:
        ImportError: If cedar-policy-mcp-schema-generator is not installed.
        RuntimeError: If schema generation fails.
    """
    try:
        from cedar_mcp_schema_generator import SchemaGeneratorError, generate_schema_or_raise
    except ImportError as e:
        raise ImportError(
            "cedar-policy-mcp-schema-generator is required for auto schema generation. "
            "Install it with: pip install cedar-policy-mcp-schema-generator"
        ) from e

    try:
        result = generate_schema_or_raise(
            DEFAULT_STUB, cast(list[dict[str, Any]], tools), config={"flattenNamespaces": True}
        )
    except SchemaGeneratorError as e:
        raise RuntimeError(f"Schema generation failed: {e}") from e

    schema = result["schema"]
    if not schema:
        raise RuntimeError("Schema generation returned empty schema")

    # The generator wraps output in "namespace Agent { ... }" with prefixed types (Agent::User).
    # Cedar evaluation uses a flat schema, so strip the wrapper and namespace prefixes.
    ns_match = re.match(r"^namespace\s+(\w+)\s*\{", schema)
    ns = ns_match.group(1) if ns_match else "Agent"

    stripped = re.sub(r"^namespace\s+\w+\s*\{", "", schema)
    stripped = re.sub(r"\}\s*$", "", stripped)
    stripped = stripped.replace(f"{ns}::", "")

    return stripped

File: strands-ts/src/vended-interventions/cedar/schema-generator.ts
import type { ToolDefinition } from './cedar.js'

/** Adapter for `@cedar-policy/mcp-schema-generator-wasm` — generates Cedar schemas from tool definitions. */
export interface SchemaGenerator {
  generateSchema(tools: ToolDefinition[]): string
}

/** Creates a SchemaGenerator from the loaded `@cedar-policy/mcp-schema-generator-wasm` module. */
export function createSchemaGenerator(
  wasm: {
    generateSchema: (stub: string, toolsJson: string, configJson?: string) => string
  },
  namespace?: string
): SchemaGenerator {
  const ns = namespace ?? 'Agent'
  const defaultStub = `
namespace ${ns} {
  @mcp_principal
  entity User;
  @mcp_resource
  entity Resource;
  @mcp_context("session")
  type SessionContext = {
    hour_utc: Long,
    call_count: Long
  };
}
`

  return {
    generateSchema(tools: ToolDefinition[]): string {
      const config = JSON.stringify({ flattenNamespaces: true })
      const result = JSON.parse(wasm.generateSchema(defaultStub, JSON.stringify(tools), config)) as {
        schema: string | null
        error: string | null
        isOk: boolean
      }
      if (!result.isOk || !result.schema) {
        throw new Error(`Schema generation failed: ${result.error}`)
      }
      // When namespace is configured, preserve the schema as-is (namespaced).
      // Otherwise, strip the namespace wrapper for backward compatibility.
      if (namespace) {
        return result.schema
      }
      const nsMatch = result.schema.match(/^namespace\s+(\w+)\s*\{/)
      const ns = nsMatch ? nsMatch[1]! : 'Agent'
      return result.schema
        .replace(/^namespace\s+\w+\s*\{/, '')
        .replace(/\}\s*$/, '')
        .replaceAll(`${ns}::`, '')
    },
  }
}
